### DISCUSSION TITLE
How to efficiently find the first matching regex from a (long) list of regexes?

### DISCUSSION THREAD

**Discussion by psychon** [rust-lang/regex#1138] [Q&A] [answered]
Hi,

imagine a poor man's translation service. The input is a message from a well-known set of messages and the output should be the same message in another language. However, the input text can contain numbers. A (possibly stupid) idea is to extract these numbers with a regex.

Random example: `There are 5 items in your shopping cart` matches `There are (\d+) items in your shopping cart` and the capture group extracts the number so that it can be inserted into another message.

Now, in practice there can be many different messages that need to be translated, so the task is to find the first regex out of a long list that matches the input. Just iterating over these can be slow and I feel like there should be faster approaches for this. (And "just iterating" likely is fast enough anyway; I'm just curious about better solutions.)

Thanks!

P.S.: Yes, I know that this is not how one is supposed to implement localisation.

**Comment by BurntSushi** [comment, marked as answer]
Does [`RegexSet`](https://docs.rs/regex/latest/regex/struct.RegexSet.html) work for you? Note its limitations however.

One thing that would make this question easier to answer is if you provided a code example implementing what you describe in prose. Something simple that captures the essence of what you're trying to do. But is ultimately something with some inputs and can be run by others. Then I could see if it makes sense to transform it to `RegexSet` or possibly something else entirely.

**Reply by psychon** [reply]
Thanks a lot! That's exactly what I was looking for, but for some reason I insisted that this must be a list and never thought about sets and so totally missed that. To my shame, I feel like I even saw this docs page some months ago and forgot about it again.

**Reply by BurntSushi** [reply]
Hah no worries.

Another API to consider is the meta regex engine in the lower level `regex-automata` crate: https://docs.rs/regex-automata/latest/regex_automata/meta/index.html

(_Both_ `Regex` and `RegexSet` are actually light wrappers around the meta engine.)
